tests/loops/test_loops.py

```diff
@@ -111,6 +111,7 @@ class Simple(Loop):
 
         def __init__(self, dataset: Iterator):
             super().__init__()
+            self.iteration_count = 0
```

Comment:
This is the case I was talking about above: If we add some dummy progress here, it is much easier for users to use that. Otherwise they have to overwrite hooks just for progress tracking (which would be kind of boilerplate)

Reply:
@carmocca this is what I meant the other day and when I accurately predicted that Justus will say the same 😄

Reply:
The fact that none of our loops (after implementing tracking) use the iteration count tells me that it's unwanted boilerplate

I would agree to keep it if it was optional (as in, could be enabled-disabled), but forcing this attribute on all loops subclasses doesn't seem right to me and can be confusing for users if they expect it to be correctly updated.

Additionally, it doesn't allow setting a semantic meaning for it (your epoch loops wants the iteration count to be named epoch count) and also doesn't easily allow doing a custom increase pattern (`*= 2` for example) without overriding `run()`.

If you really want to provide a loop with iteration count, then, since setting an attribute optionally is ugly and error prone we add one Loop class (like `SimpleLoop`) which does add the `iteration_count`.

This would be similar to our `AbstractProfiler`/`BaseProfiler` distinction.

Reply:
Also note that `iteration_count` wasn't already invisible to this concrete before this PR.

It still needs to be used in the `done` implementation and the `reset` implementation. So the users will need some sort of control over it anyways.

Reply:
> The fact that none of our loops (after implementing tracking) use the iteration count tells me that it's unwanted boilerplate

They don't have it any more because tracking replaced it. So if a loop has tracking, it's unwanted boilerplate, I agree. But if a loop doesn't have tracking (like the base loop), I don't think it's boilerplate.

> Additionally, it doesn't allow setting a semantic meaning for it (your epoch loops wants the iteration count to be named epoch count) and also doesn't easily allow doing a custom increase pattern (*= 2 for example) without overriding run().

That's right. So I'm fine with the solution you propose (Adding SimpleLoop). There we maybe can also have some properties, that define the missing behaviour there.

Reply:
I agree with Justus there. I believe Lightning should provide the iteration_counter by default for loop customisation, but people can de-activate it when they have added progress properly.

```py
class Loop

    def __init__(self, use_iteration_counter: bool = True)
        self._use_iteration_counter = use_iteration_counter

# for users without tracking
class MyLoop(Loop):

    def __init__(...):
        super()


# for users with tracking
class MyLoop(Loop):

    def __init__(...):
        super(use_iteration_counter=False)
```

Reply:
> but people can de-activate it when they have added progress properly.

I don't like the flag solution as now you cannot rely on the attribute being present as it would depend on the flag.

In any case, this can be a follow-up PR when loop customization is closer.